File: examples/calculator/js/serve.py
#!/usr/bin/env python3
"""Serve the calculator browser example with Emscripten's required headers."""

import argparse
import functools
import mimetypes
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path


class CalculatorRequestHandler(SimpleHTTPRequestHandler):
    """Static file handler configured for the pthread-enabled wasm module."""

    extensions_map = {
        **mimetypes.types_map,
        ".mjs": "text/javascript",
        ".wasm": "application/wasm",
    }

    def end_headers(self):
        self.send_header("Cross-Origin-Opener-Policy", "same-origin")
        self.send_header("Cross-Origin-Embedder-Policy", "require-corp")
        super().end_headers()


def parse_arguments():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--directory",
        type=Path,
        default=Path(__file__).absolute().parent,
        help="directory containing index.html and the generated module",
    )
    parser.add_argument("--host", default="127.0.0.1")
    parser.add_argument("--port", type=int, default=8000)
    return parser.parse_args()


def main():
    arguments = parse_arguments()
    directory = arguments.directory.expanduser().resolve()
    handler = functools.partial(CalculatorRequestHandler, directory=str(directory))
    server = ThreadingHTTPServer((arguments.host, arguments.port), handler)
    print(f"Serving {directory} at http://{arguments.host}:{arguments.port}/index.html")
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\nServer stopped")
    finally:
        server.server_close()


if __name__ == "__main__":
    main()

File: docs/js_binding_dev/spikes/pthreads_callbacks_spike/serve.mjs
// Minimal static file server with COOP/COEP headers so that the pthread build
// can use SharedArrayBuffer in a browser. Run from this directory:
//
//   node serve.mjs [port]
import { createServer } from "node:http";
import { readFile } from "node:fs/promises";
import { extname, join } from "node:path";
import { fileURLToPath } from "node:url";

const root = fileURLToPath(new URL(".", import.meta.url));
const port = Number(process.argv[2]) || 8080;

const contentTypes = {
    ".html": "text/html",
    ".mjs": "text/javascript",
    ".js": "text/javascript",
    ".wasm": "application/wasm",
};

createServer(async (request, response) => {
    const path = request.url === "/" ? "/test.html" : request.url.split("?")[0];
    try {
        const body = await readFile(join(root, path));
        response.writeHead(200, {
            "Content-Type": contentTypes[extname(path)] ?? "application/octet-stream",
            "Cross-Origin-Opener-Policy": "same-origin",
            "Cross-Origin-Embedder-Policy": "require-corp",
        });
        response.end(body);
    } catch {
        response.writeHead(404).end();
    }
}).listen(port, () => console.log(`serving ${root} at http://localhost:${port}`));
